src/Microsoft.AspNet.Mvc.Core/ActionResults/ViewExecutor.cs

```diff
@@ -23,22 +29,44 @@ public static class ViewExecutor
         /// <param name="actionContext">The <see cref="ActionContext"/> for the current executing action.</param>
         /// <param name="viewData">The <see cref="ViewDataDictionary"/> for the view being rendered.</param>
         /// <param name="tempData">The <see cref="ITempDataDictionary"/> for the view being rendered.</param>
-        /// <returns>A <see cref="Task"/> that represents the asychronous rendering.</returns>
+        /// <returns>A <see cref="Task"/> that represents the asynchronous rendering.</returns>
         public static async Task ExecuteAsync([NotNull] IView view,
                                               [NotNull] ActionContext actionContext,
                                               [NotNull] ViewDataDictionary viewData,
                                               [NotNull] ITempDataDictionary tempData,
-                                              string contentType)
+                                              MediaTypeHeaderValue contentType)
         {
-            if (string.IsNullOrEmpty(contentType))
+            var response = actionContext.HttpContext.Response;
+
+            var contentTypeHeader = contentType;
+            Encoding encoding;
+            if (contentTypeHeader == null)
             {
-                contentType = ContentType;
+                contentTypeHeader = DefaultContentType;
+                encoding = DefaultContentType.Encoding;
+            }
+            else
+            {
+                if (contentTypeHeader.Encoding == null)
+                {
+                    // 1. Do not modify the user supplied content type
+                    // 2. Parse here to handle parameters apart from charset
+                    contentTypeHeader = MediaTypeHeaderValue.Parse(contentTypeHeader.ToString());
+                    contentTypeHeader.Encoding = Encoding.UTF8;
+                }
+
+                encoding = contentTypeHeader.Encoding;
             }
 
-            actionContext.HttpContext.Response.ContentType = contentType;
-            var wrappedStream = new StreamWrapper(actionContext.HttpContext.Response.Body);
-            var encoding = Encodings.UTF8EncodingWithoutBOM;
-            using (var writer = new StreamWriter(wrappedStream, encoding, BufferSize, leaveOpen: true))
+            response.ContentType = contentTypeHeader.ToString();
+
+            var wrappedStream = new StreamWrapper(response.Body);
+
+            using (var writer = new StreamWriter(
```

Comment:
Could you add a comment here calling out why we have a wrapper for the encoding?